src/main/java/thecodex6824/thaumicaugmentation/common/item/ItemImpulseCannonAugment.java

```diff
@@ -20,187 +20,293 @@
-                    PacketRecoil recoil = new PacketRecoil(user.getEntityId(), RecoilType.IMPULSE_BURST);
-                    TANetwork.INSTANCE.sendToAllTracking(recoil, user);
-                    if (user instanceof EntityPlayer) {
-                        ((EntityPlayer) user).getCooldownTracker().setCooldown(TAItems.IMPULSE_CANNON, TAConfig.cannonBurstCooldown.getValue());
-                        if (user instanceof EntityPlayerMP)
-                            TANetwork.INSTANCE.sendTo(recoil, (EntityPlayerMP) user);
-                    }
+            }
+
+
+            @Override
+            public void applyAdditionalEffects(EntityLivingBase user, Vec3d firingOrigin, Vec3d firingEnd, float baseDamage) {
+                Vec3d offset = firingEnd.subtract(firingOrigin);
+                firingEnd = offset.add(offset.normalize().scale(Math.sqrt(baseDamage))).add(firingOrigin);
+                int tick = FMLCommonHandler.instance().getMinecraftServerInstance().getTickCounter();
+                while (baseDamage > 0) {
+                    RayTraceResult result = user.getEntityWorld().rayTraceBlocks(firingOrigin, firingEnd, false, true, false);
+                    if (result == null || result.typeOfHit != RayTraceResult.Type.BLOCK) break;
+                    baseDamage = BreakInformation.damageBlock(user.getEntityWorld(), result.getBlockPos(), baseDamage, tick);
                 }
-                
-            };
-        }
+            }
+        };
+        // purifier
+        augments[5] = new IImpulseCannonAugment() {
+
+            @Override
+            public double getImpulseCostModifier(IImpetusStorage buffer) {
+                return 2;
+            }
+
+            @Override
+            public float getBaseDamageModifier(IImpetusStorage buffer, double normalImpetusConsumed, double actualImpetusConsumed) {
+                return 0.2f;
+            }
+
+            @Override
+            public void applyAdditionalEffectsToEntity(EntityLivingBase user, Vec3d firingOrigin, Vec3d firingEnd, Entity entityHit, float baseDamage) {
+                DamageSource source = new DamageSourceImpetus(user, user.getPositionVector()).setDamageBypassesArmor().setDamageIsAbsolute();
+                entityHit.attackEntityFrom(source, baseDamage * 2);
+                if (entityHit instanceof EntityLivingBase base) {
+                    base.hurtResistantTime = Math.min(base.hurtResistantTime, 2);
+                    base.lastDamage = 0.0F;
+                }
+            }
+        };
     }
     
     @Override
     @Nullable
     public ICapabilityProvider initCapabilities(ItemStack stack, @Nullable NBTTagCompound nbt) {
         SimpleCapabilityProviderNoSave<IAugment> provider =
-                new SimpleCapabilityProviderNoSave<>(createAugmentForStack(stack), CapabilityAugment.AUGMENT);
+                new SimpleCapabilityProviderNoSave<>(augments[stack.getMetadata()], CapabilityAugment.AUGMENT);
```

Comment:
This will put the same augment instance into every stack of the same type, which is fine now but would be a problem if the augments ever started to contain state. Unless these augment instances are extremely expensive to create, I think it'd be better to just make new ones to avoid leaving the shared instance footgun here.